### PR TITLE
Fix the WASM Worker cannot start in node.js environment

### PR DESCRIPTION
**Pull request by medns**
`emscripten_malloc_wasm_worker` and
`emscripten_create_wasm_worker` functions
adapted to the nodejs environment and
modified the core related tests.

### CODE REVIEW

**Comment by medns** [comment]
`ci/circleci: test-wasm64` exceeds the maximum execution time and may need to be rerun. 
All my local tests have passed.

File: src/library_wasm_worker.js [hunk 1]
```diff
@@ -47,7 +47,12 @@ addToLibrary({
   $_wasmWorkerRunPostMessage: (e) => {
     // '_wsc' is short for 'wasm call', trying to use an identifier name that
     // will never conflict with user code
-    let data = e.data, wasmCall = data['_wsc'];
+#if ENVIRONMENT_MAY_BE_NODE
+    const data = ENVIRONMENT_IS_NODE ? e : e.data;
+#else
+    const data = e.data;
+#endif
+    const wasmCall = data['_wsc'];
```
Comment (sbc100) [inline src/library_wasm_worker.js:55]: Using `const` instead of `let` regresses code size here.   Maybe revert to `const`?

File: test/test_core.py [hunk 2]
```diff
@@ -9795,15 +9795,15 @@ def test_emscripten_async_load_script(self):
 
   @node_pthreads
   def test_wasm_worker_hello(self):
-    self.do_runf(test_file('wasm_worker/hello_wasm_worker.c'), emcc_args=['-sWASM_WORKERS'])
+    self.do_run_in_out_file_test(test_file('wasm_worker/hello_wasm_worker.c'), emcc_args=['-sWASM_WORKERS'], timeout=3, timeout_as_error=False)
```
Comment (sbc100) [inline test/test_core.py:9798]: Why the timeout here?

**Review by sbc100 — COMMENTED** [review]
I'll let @juj do the main review here.

File: src/library_wasm_worker.js (hunk 1 above)
Comment (medns) [reply src/library_wasm_worker.js]: reverted

File: test/test_core.py (hunk 2 above)
Comment (medns) [reply test/test_core.py:9798]: Because the Node.js process will not exit automatically after the WASM Worker finishes running in Node.js, the `run_js` method in the unit test jsrun.py will not return (blocking and waiting for the process under test to exit).
So the `timeout=3` + `timeout_as_error=False` method is used here. When the timeout period is exceeded, the `run_js` method in jsrun.py will return, so the correct stdout information can be obtained. For the `hello wasm worker` test, 3 seconds is enough time to complete the run.
If do not use this method, the test method in the unit test needs to be changed to stream comparison, and the overall modification cost is relatively high.

Comment (medns) [reply test/test_core.py:9798]: Although it is not optimal and is just a trick, but it can solve the current PR testing problem at the lowest cost.

Comment (sbc100) [reply test/test_core.py:9798]: I think it might be better to instead inject some kind of `--pre-js` file.. that does `setTimeout(..., 3)` to trigger node to exit, rather than adding complexity to the test runner.

Also can you add TODO to fix the exit handling in the wasm workers case?

Comment (medns) [reply test/test_core.py]: The `--pre-js` method has been used to trigger the exit of nodejs.

File: test/test_core.py [hunk 3]
```diff
@@ -9793,16 +9793,23 @@ def test_emscripten_async_load_script(self):
     self.run_process([FILE_PACKAGER, 'test.data', '--preload', 'file1.txt', 'file2.txt', '--from-emcc', '--js-output=script2.js'])
     self.do_runf(test_file('test_emscripten_async_load_script.c'), emcc_args=['-sFORCE_FILESYSTEM'])
 
+  def prep_wsam_worker_in_node(self):
```
Comment (sbc100) [inline test/test_core.py:9796]: wsam -> wasm

File: test/test_core.py [hunk 4]
```diff
@@ -9793,16 +9793,23 @@ def test_emscripten_async_load_script(self):
...
+    # Auto exit after 3 seconds in Nodejs environment to get WASM Worker stdout
```
Comment (sbc100) [inline test/test_core.py:9797]: maybe add `TODO: remove this is if/when wasm worker tests exit cleanly under node`

File: src/preamble.js [hunk 5]
```diff
@@ -246,6 +242,13 @@ function initRuntime() {
   callRuntimeCallbacks(__RELOC_FUNCS__);
 #endif
   <<< ATINITS >>>
+  
+#if WASM_WORKERS
+if (ENVIRONMENT_IS_WASM_WORKER) {
+  return _wasmWorkerInitializeRuntime();
+}
+#endif
```
Comment (juj) [inline src/preamble.js:250]: This change does not look correct. `_wasmWorkerInitializeRuntime()` initializes the stack and TLS of the Worker, which is to be done instead of `checkStackCookie();` and `setStackLimits();` above.

It loks like after this code, the stack would be initialized twice, and `__RELOC_FUNCS__` and `<<< ATINITS >>>` would also be run in Wasm Workers, but they aren't run in Pthreads either?

What was the motivation for this change?

File: src/wasm_worker.js [hunk 6]
```diff
@@ -1,18 +1,51 @@
...
+    importScripts: (f) => (0, eval)(fs.readFileSync(f, 'utf8') + '//# sourceURL=' + f),
+    postMessage: (msg) => parentPort.postMessage(msg),
+    performance: global.performance || { now: Date.now },
+    addEventListener: (name, handler) => parentPort.on(name, handler),
+    removeEventListener: (name, handler) => parentPort.off(name, handler),
+  });
+}
+#endif // ENVIRONMENT_MAY_BE_NODE
+
+self.onmessage = function(d) {
```
Comment (juj) [inline src/wasm_worker.js:37]: Would it be a bother to add to src/parseTools.js

```js
function implicitSelf() {
  return ENVIRONMENT.includes('node') ? 'self.' : '';
}
```
and do
```
{{{implicitSelf()}}}onmessage = ...;
```
to keep the code size intact?

Or alternatively a

```js
#if ENVIRONMENT_MAY_BE_NODE
self.
#endif
  onmessage = ...;
```
construct?

File: src/preamble.js (hunk 5 above)
Comment (medns) [reply src/preamble.js:250]: Since `_wasmWorkerInitializeRuntime()` returns early before `<<< ATINITS >>>`, cause `FS.init` in <<< ATINITS >>> is not been called, resulting in stdin(0), stdout(1) and stderr(2) these three device is not been created in WASM Worker environment.
When the code uses pipes for output (such as printf or cout), will throw a device dose not exists error. Especially when the AddressSanitizer tool is enabled by `-fsanitize=address` flag.
So I modified the order to ensure `<<< ATINITS >>>` can be executed.

File: src/wasm_worker.js (hunk 6 above)
Comment (medns) [reply src/wasm_worker.js:37]: Okey

File: src/preamble.js (hunk 5 above)
Comment (medns) [reply src/preamble.js:261]: I adjusted the calling order so that now only the methods in `_wasmWorkerInitializeRuntime` and `<<< ATINITS >>>` will be called in WASM Worker environment.

**Comment by medns** [comment]
Failed tests in `ci/circleci: test-core2` and `ci/circleci: test-wasm2js1` may need to be rerun.
All my local tests have been passed, as follows:
image

File: test/test_core.py (hunk 3 above)
Comment (kripken) [reply test/test_core.py:9841]: This typo is still here it seems?

File: test/test_core.py (hunk 4 above)
Comment (kripken) [reply test/test_core.py:9842]: this TODO also still needs to be added.

File: src/wasm_worker.js [hunk 7]
```diff
@@ -1,18 +1,51 @@
...
+    importScripts: (f) => (0, eval)(fs.readFileSync(f, 'utf8') + '//# sourceURL=' + f),
+    postMessage: (msg) => parentPort.postMessage(msg),
+    performance: global.performance || { now: Date.now },
+    addEventListener: (name, handler) => parentPort.on(name, handler),
+    removeEventListener: (name, handler) => parentPort.off(name, handler),
+  });
+}
+#endif // ENVIRONMENT_MAY_BE_NODE
+
+{{{implicitSelf()}}}onmessage = function(d) {
```
Comment (kripken) [inline src/wasm_worker.js:37]:
```suggestion
{{{ implicitSelf() }}}onmessage = function(d) {
```

File: src/wasm_worker.js [hunk 8]
```diff
@@ -1,18 +1,51 @@
...
 	// The first message sent to the Worker is always the bootstrap message.
 	// Drop this message listener, it served its purpose of bootstrapping
 	// the Wasm Module load, and is no longer needed. Let user code register
 	// any desired message handlers from now on.
-	onmessage = null;
+	{{{implicitSelf()}}}onmessage = null;
```
Comment (kripken) [inline src/wasm_worker.js:42]:
```suggestion
	{{{ implicitSelf() }}}onmessage = null;
```

File: src/preamble.js (hunk 5 above)
Comment (juj) [reply src/preamble.js:261]: The semantics of <<< ATINITS >>> is to not be called in startup of pthreads either (note the if (ENVIRONMENT_IS_PTHREAD) return; block), so the semantics should equally be preserved to not be called at startup of wasm workers as well.

Some people might have ATINITS that should only run once at whole program startup, so this would change the semantics on them that they'd be inadvertently rerunning whenever a new wasm worker is created, which would break their programs.

I think we may need a custom solution for the wasm worker stdio situation, maybe something that parallels pthreads. I forget, how did pthreads get to initialize the stdio handles so printing works there?

Comment (sbc100) [reply src/preamble.js:261]: IIUC, with pthreads all FS calls are proxied back to the main thread.  i.e. None of FS API is designed to be called from a worker.

Comment (juj) [reply src/preamble.js:261]: Oh right..

I wonder if using WASMFS avoids all of that to make it work without hassle?

Maybe we'd want to do a separate `<<< WORKER_ATINITS >>>` type of section that would run for each Worker startup?

There library_fs.js could run the needed stdio setup code that is desirable for each Worker context, and in this case gate it behind `if (ENVIRONMENT_IS_WASM_WORKER)`. (That kind of mechanism might be more generic than crafting a Wasm Worker specific <<< WASM_WORKER_ATINITS >>> block)

Comment (sbc100) [reply src/preamble.js:261]: Are the filesystem APIs expected to work in wasm worker though?    I don't really see how they could.  Shouldn't try assert/abort because proxying is not available?

Comment (medns) [reply src/preamble.js]: Hey, guys,
I found an alternative solutions to solve this problem. 
Instead of modifying the init script of the WASM worker, I adjusted the implementation of `fd_write(...)` method in `library_wasi.js`.
When the code runs inside the WASM worker, use a simple way to support stdout and stderr pipes.

File: src/library_wasi.js [hunk 9]
```diff
@@ -261,17 +260,41 @@ var WasiLibrary = {
     if (printCharBuffers[1].length) printChar(1, {{{ charCode("\n") }}});
     if (printCharBuffers[2].length) printChar(2, {{{ charCode("\n") }}});
   },
-  fd_write__deps: ['$flush_NO_FILESYSTEM', '$printChar'],
-  fd_write__postset: () => addAtExit('flush_NO_FILESYSTEM()'),
+#if SYSCALLS_REQUIRE_FILESYSTEM
+  fd_write__postset: () => addAtExit('if(ENVIRONMENT_IS_WASM_WORKER) flush_NO_FILESYSTEM()'),
```
Comment (sbc100) [inline src/library_wasi.js:264]: I don't think this is needed is the `addAtExit` mechanism is already limited to just the main thread.

File: src/parseTools.js [hunk 10]
```diff
@@ -1023,3 +1035,7 @@ function getPerformanceNow() {
     return 'performance.now';
   }
 }
+
+function implicitSelf() {
+  return ENVIRONMENT.includes('node') ? 'self.' : '';
```
Comment (sbc100) [inline src/parseTools.js:1040]: What happens if the environment includes both node and web and we run the result on the web?    i.e. isn't choosing to use `self.` or not here at compile time too early.. don't we need to decide at runtime maybe?

File: src/preamble.js [hunk 11]
```diff
@@ -786,7 +786,7 @@ function instantiateSync(file, info) {
 }
 #endif
 
-#if PTHREADS && (LOAD_SOURCE_MAP || USE_OFFSET_CONVERTER)
+#if (PTHREADS || WASM_WORKERS) && (LOAD_SOURCE_MAP || USE_OFFSET_CONVERTER)
```
Comment (sbc100) [inline src/preamble.js:789]: Maybe use `SHARED_MEMORY` as a shorthand for `PTHREADS || WASM_WORKERS`?

File: src/library_wasi.js [hunk 12]
```diff
@@ -261,17 +260,41 @@ var WasiLibrary = {
...
+#if WASM_WORKERS
+  } else {
+    // Hack to support printf in WASM workers in a simple way.
+    // We support at least stdout, stderr.
+#endif // WASM_WORKERS
+#endif // SYSCALLS_REQUIRE_FILESYSTEM
+#if !SYSCALLS_REQUIRE_FILESYSTEM || WASM_WORKERS
+#if !SYSCALLS_REQUIRE_FILESYSTEM
+    // Hack to support printf in SYSCALLS_REQUIRE_FILESYSTEM=0
+#endif
```
Comment (sbc100) [inline src/library_wasi.js:297]: Kind of shame that we now have to support the hybrid approach of both real FS stuff and fake FS stuff in the same build.  

To avoid some of these confusing `#if` maybe we could split out the simple/dummy version fs_write and give is its own name  (e.g. `fdWriteNOFS`)

Then we could generate something like:

```
fd_write: (fd, iov, iovcnt, pnum) => {
   if (ENVIRONMENT_IS_WASM_WORKER) {
      return fdWriteNOFS(...)
   }
   ... nomral FS code ..
}
```

And in the non-worker case where no filesystem is needed you could use an alias:

```
#if !SYSCALLS_REQUIRE_FILESYSTEM
fd_write: '$fdWriteNOFS',`
#endif
```

File: src/parseTools.js (hunk 10 above)
Comment (medns) [reply src/parseTools.js:1040]: In the web environment, it can run with or without the `self.` prefix, but in the node environment, it must have the `self.` prefix to run.
So I think the above code is correct.

Comment (sbc100) [reply src/parseTools.js:1040]: Ah I see.  It looks like we have the same pattern in `src/worker.js` but we simply always use `self` there.

I wonder if we could find a way to avoid the `self.` completely by using `defineProperty` on the global object under node?  Maybe that would be more code overall?

File: src/wasm_worker.js [hunk 13]
```diff
@@ -1,18 +1,51 @@
...
+    },
+    Worker: nodeWorkerThreads.Worker,
+    importScripts: (f) => (0, eval)(fs.readFileSync(f, 'utf8') + '//# sourceURL=' + f),
+    postMessage: (msg) => parentPort.postMessage(msg),
+    performance: global.performance || { now: Date.now },
+    addEventListener: (name, handler) => parentPort.on(name, handler),
+    removeEventListener: (name, handler) => parentPort.off(name, handler),
+  });
+}
+#endif // ENVIRONMENT_MAY_BE_NODE
```
Comment (sbc100) [inline src/wasm_worker.js:35]: This seems like duplication of the code in `src/worker.js` .. is it worth factoring out into a separate file that can be included in both places.  Something like `src/node_webworker_polyfill.py`?

Comment (medns) [reply src/wasm_worker.js:35]: I think this issue is the same as the [issue](https://github.com/emscripten-core/emscripten/pull/20188#discussion_r1338807286) above, can create a separate PR after merging this to deal with these two issues together.

File: src/parseTools.js (hunk 10 above)
Comment (medns) [reply src/parseTools.js:1040]: I think this issue is the same as the [issue](https://github.com/emscripten-core/emscripten/pull/20188#discussion_r1338809434) below, it can create a separate PR after merging this to deal with these two issues together.

File: src/library_wasi.js [hunk 14]
```diff
@@ -282,10 +274,30 @@ var WasiLibrary = {
...
+  fd_write__deps: [
+    '$doWritev',
+#if WASM_WORKERS
+    '$fd_write_nofs',
+#endif
+  ],
+  fd_write: (fd, iov, iovcnt, pnum) => {
+#if WASM_WORKERS
+    if (ENVIRONMENT_IS_WASM_WORKER) {
+      return fd_write_nofs(fd, iov, iovcnt, pnum);
```
Comment (sbc100) [inline src/library_wasi.js:290]: @juj @kripken is this the behaviour we want?

If the application is setup to use the real filesystem, we would end up with two different filesystem behaviours, one in the wasm workers (which would see a dummy/fake filesystem just for stdout) and a different behaviour on the main thread and on the pthreads?

Comment (sbc100) [reply src/library_wasi.js:290]: An alternative to doing this this way would be to monkey-patch all the filesystem functions on wasm worker startup so it sees the dummy/fake ones instead of the fully-functional ones?

File: src/library_wasi.js (hunk 12 above)
Comment (medns) [reply src/library_wasi.js:265]: Modified as required.

File: test/test_core.py (hunk 4 above)
Comment (medns) [reply test/test_core.py:9842]: Modified as required

File: src/library_wasi.js (hunk 14 above)
Comment (sbc100) [reply src/library_wasi.js:292]: I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: https://github.com/emscripten-core/emscripten/pull/20343

Comment (medns) [reply src/library_wasi.js:293]: > @juj @kripken is this the behaviour we want?
> 
> If the application is setup to use the real filesystem, we would end up with two different filesystem behaviours, one in the wasm workers (which would see a dummy/fake filesystem just for stdout) and a different behaviour on the main thread and on the pthreads?

According to the previous discussion, there is currently no available filesystem in the WASM workers, so I think the behavior in the WASM workers should be consistent with `-sFILESYSTEM=0` (no filesystem), and this part of the logic can be reused.

Comment (sbc100) [reply src/library_wasi.js:293]: I guess that make some sense, but it also make the code larger because we are including both versions of the FS code at the same time.

Another approach might be simply to make all FS syscalls fail on wasm workers (like most other syscalls do). Users could still call `emscripten_console_log` and `emscripten_console_logf` to do write to the console but the wouldn't be able to use normal stdio.  Given that wasm workers are already massively restricted in what they can do perhaps this is a reasonable restriction?

Comment (medns) [reply src/library_wasi.js:293]: > I guess that make some sense, but it also make the code larger because we are including both versions of the FS code at the same time.
> 
> Another approach might be simply to make all FS syscalls fail on wasm workers (like most other syscalls do). Users could still call `emscripten_console_log` and `emscripten_console_logf` to do write to the console but the wouldn't be able to use normal stdio. Given that wasm workers are already massively restricted in what they can do perhaps this is a reasonable restriction?

I considered this option, but due to the widespread use of normal stdio it would be a big breaking change to disable this behavior in the WASM worker.

Comment (sbc100) [reply src/library_wasi.js:293]: @juj WDTY?

Comment (juj) [reply src/library_wasi.js:293]: > I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343

Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.

The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.

But we do want #20343 more, because it will be able to identify unsupported things better.

I kind of wish we could start the `printf()` function implementation with a 

```c++
void printf(...) {
  if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
  else printf_regular(...);
}
```
so that we'd route the implementation away from any filesystem in these types of scenarios.

@medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

Comment (sbc100) [reply src/library_wasi.js:293]: Converting the tests from using `printf` to `emscripten_console_logf` would probably be a good idea (at least for now).

Comment (medns) [reply src/library_wasi.js:293]: > > I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343
> 
> Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.
> 
> The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.
> 
> But we do want #20343 more, because it will be able to identify unsupported things better.
> 
> I kind of wish we could start the `printf()` function implementation with a
> 
> ```c++
> void printf(...) {
>   if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
>   else printf_regular(...);
> }
> ```
> 
> so that we'd route the implementation away from any filesystem in these types of scenarios.
> 
> @medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

> > I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343
> 
> Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.
> 
> The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.
> 
> But we do want #20343 more, because it will be able to identify unsupported things better.
> 
> I kind of wish we could start the `printf()` function implementation with a
> 
> ```c++
> void printf(...) {
>   if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
>   else printf_regular(...);
> }
> ```
> 
> so that we'd route the implementation away from any filesystem in these types of scenarios.
> 
> @medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

Okay, let me strip out the modifications about filesystem and use `emscripten_console_logf` instead of `printf` to implement unit tests.
I'm on vacation for the next few days and will submit new patch in early October.

Comment (medns) [reply src/library_wasi.js]: Modified as required.

File: src/preamble.js [hunk 15]
```diff
@@ -1082,22 +1082,18 @@ function createWasm() {
   // path.
   if (Module['instantiateWasm']) {
 
-#if USE_OFFSET_CONVERTER && PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) {
+#if USE_OFFSET_CONVERTER
 #if ASSERTIONS
-      assert(Module['wasmOffsetData'], 'wasmOffsetData not found on Module object');
+{{{ runIfWorkerThread("assert(Module['wasmOffsetData'], 'wasmOffsetData not found on Module object');") }}}
```
Comment (sbc100) [inline src/preamble.js:1087]: I think this would perhaps be move readable if just the precondition was a macro.

e.g.:
```
if ({{{ runningInWorker() }}}) {
   assert(Module['wasmOffsetData'], 'wasmOffsetData not found on Module object');") }}}
}
```

That way we not writing JS code inside the macro itself.  WDTY?

**Review by sbc100 — COMMENTED** [review]
lgtm % comments from me.  I'll leave @juj to approve

File: src/preamble.js (hunk 15 above)
Comment (medns) [reply src/preamble.js:1087]: > I think this would perhaps be move readable if just the precondition was a macro.
> 
> e.g.:
> 
> ```
> if ({{{ runningInWorker() }}}) {
>    assert(Module['wasmOffsetData'], 'wasmOffsetData not found on Module object');") }}}
> }
> ```
> 
> That way we not writing JS code inside the macro itself. WDTY?

`runningInWorker` mimics `runningInMain`, as shown below:

https://github.com/emscripten-core/emscripten/blob/e18753e6fbbf2d690e5a750165c693bb6bfec122/src/preamble.js#L735

Comment (sbc100) [reply src/preamble.js:1087]: Fair enough.

File: src/parseTools.js [hunk 16]
```diff
@@ -736,6 +736,18 @@ function runIfMainThread(text) {
   }
 }
 
+function runIfWorkerThread(text) {
+  if (WASM_WORKERS && PTHREADS) {
+    return 'if (ENVIRONMENT_IS_WASM_WORKER || ENVIRONMENT_IS_PTHREAD) { ' + text + ' }';
+  } else if (WASM_WORKERS) {
+    return 'if (ENVIRONMENT_IS_WASM_WORKER) { ' + text + ' }';
+  } else if (PTHREADS) {
+    return 'if (ENVIRONMENT_IS_PTHREAD) { ' + text + ' }';
```
[the remark this answers is not in the document]
Comment (medns) [reply src/parseTools.js:745]: Okey

File: src/library_wasm_worker.js [hunk 17]
```diff
@@ -47,7 +47,12 @@ addToLibrary({
   $_wasmWorkerRunPostMessage: (e) => {
     // '_wsc' is short for 'wasm call', trying to use an identifier name that
     // will never conflict with user code
-    let data = e.data, wasmCall = data['_wsc'];
+#if ENVIRONMENT_MAY_BE_NODE
+    let data = ENVIRONMENT_IS_NODE ? e : e.data;
+#else
+    let data = e.data;
+#endif
```
Comment (kripken) [inline src/library_wasm_worker.js:54]: Why is node different here?

Please add a comment here, as otherwise this looks surprising to the reader.

Comment (medns) [reply src/library_wasm_worker.js:56]: Modified.

File: src/preamble.js [hunk 18]
```diff
@@ -1034,7 +1034,7 @@ function createWasm() {
     exportWasmSymbols(wasmExports);
 #endif
 
-#if PTHREADS || WASM_WORKERS
+#if SHARED_MEMORY
```
Comment (juj) [inline src/preamble.js:1037]: This change, and the change above are not strictly correct under the meaning of `SHARED_MEMORY`, but it should not be an issue either (will just increase code size for users of `-sSHARED_MEMORY=1`)

There are three different ways to target multithreading:

- `-sPTHREADS=1`, which implies `SHARED_MEMORY=1`
- `-sWASM_WORKERS=1`, which implies `SHARED_MEMORY=1`
- `-sSHARED_MEMORY=1`, which is used in situations where Emscripten compiled build should run single-threaded but using a shared WebAssembly Memory, so that developers can develop custom (non-Emscripten-compiled) Workers

In other words, `#if SHARED_MEMORY` does not mean "using pthreads or using wasm workers", but it means "using pthreads, wasm workers or manual shared memory".

The original `#if PTHREADS || WASM_WORKERS` would be more accurate here. Likewise above, the function `resetPrototype` would not be used in `-sSHARED_MEMORY=1` builds.

**Comment by juj** [comment]
Looks very nice, great work. I had one comment about the subtle distinction between SHARED_MEMORY and PTHREADS/WASM_WORKERS. Otherwise LGTM to land.

File: src/library_wasm_worker.js [hunk 19]
```diff
@@ -47,7 +47,14 @@ addToLibrary({
   $_wasmWorkerRunPostMessage: (e) => {
     // '_wsc' is short for 'wasm call', trying to use an identifier name that
     // will never conflict with user code
-    let data = e.data, wasmCall = data['_wsc'];
+#if ENVIRONMENT_MAY_BE_NODE
+    // In Node.js environment, message event 'e' containing the actual data sent,
+    // while in the browser environment it's contained by 'e.data'.
```
Comment (kripken) [inline src/library_wasm_worker.js:52]: Why is this the case, though? If it arrives from a Node.js or Web API then that means this workaround is necessary, but if not, then if it arrives from our own code, can we send data in the same format?

Comment (kripken) [reply src/library_wasm_worker.js:52]: Oh, reading that again, I think you're saying it's the first? We send data the same way, but Node and the Web end up providing it differently? If so then lgtm.

Comment (medns) [reply src/library_wasm_worker.js:52]: > Oh, reading that again, I think you're saying it's the first? We send data the same way, but Node and the Web end up providing it differently? If so then lgtm.

yeah, you are right! 👍

File: src/preamble.js (hunk 18 above)
Comment (medns) [reply src/preamble.js]: Revert to using `#if PTHREADS || WASM_WORKERS`.

**Review by kripken — APPROVED** [review]
lgtm and tests now passed.

@sbc100 did you want to take another look before we land?

**Review by sbc100 — APPROVED** [review]
I would like see the node worker stuff factored out as a followup, but otherwise lgtm
